tensorflow/lite/kernels/internal/reference/reference_ops.h

```diff
@@ -2175,39 +2175,28 @@ inline void Slice(const tflite::SliceParams& op_params,
                   const RuntimeShape& input_shape,
                   const RuntimeShape& output_shape,
                   SequentialTensorWriter<T>* writer) {
-  const RuntimeShape ext_shape = RuntimeShape::ExtendedShape(4, input_shape);
-  // TODO(dkalenichenko): This op only supports 4D tensors or smaller.
-  TFLITE_DCHECK_LE(op_params.begin_count, 4);
-  TFLITE_DCHECK_LE(op_params.size_count, 4);
-  const int begin_count = op_params.begin_count;
-  const int size_count = op_params.size_count;
-  // We front-pad the begin and size vectors.
-  const int start_b = 4 - begin_count > 0 ? 0 : op_params.begin[0];
-  const int stop_b = (4 - size_count > 0 || op_params.size[0] == -1)
-                         ? ext_shape.Dims(0)
-                         : start_b + op_params.size[0];
-  const int start_h = begin_count < 3 ? 0 : op_params.begin[begin_count - 3];
-  const int stop_h = (size_count < 3 || op_params.size[size_count - 3] == -1)
-                         ? ext_shape.Dims(1)
-                         : start_h + op_params.size[size_count - 3];
-  const int start_w = begin_count < 2 ? 0 : op_params.begin[begin_count - 2];
-  const int stop_w = (size_count < 2 || op_params.size[size_count - 2] == -1)
-                         ? ext_shape.Dims(2)
-                         : start_w + op_params.size[size_count - 2];
-  const int start_d = begin_count < 1 ? 0 : op_params.begin[begin_count - 1];
-  const int stop_d = (size_count < 1 || op_params.size[size_count - 1] == -1)
-                         ? ext_shape.Dims(3)
-                         : start_d + op_params.size[size_count - 1];
-
-  for (int in_b = start_b; in_b < stop_b; ++in_b) {
-    for (int in_h = start_h; in_h < stop_h; ++in_h) {
-      for (int in_w = start_w; in_w < stop_w; ++in_w) {
-        for (int in_d = start_d; in_d < stop_d; ++in_d) {
-          writer->Write(Offset(ext_shape, in_b, in_h, in_w, in_d));
+  const int slice_dimensions = input_shape.DimensionsCount();
+
+  // Store as std::function to allow recursion.
+  std::function<void(int, int)> compute_slice =
+      [&compute_slice, slice_dimensions, writer, input_shape, op_params](
+          int axis, int curr_indices) {
+        if (axis == (slice_dimensions - 1)) {
+          for (int in_x = op_params.begin[axis]; in_x < op_params.size[axis];
+               in_x++) {
+            int index = in_x + curr_indices;
+            writer->Write(index);
```

Comment:
I wonder why do we have to avoid WriteN here?

Reply:
@thaink : This was old code, so i did not modify the base logic. However i agree, both optimized & reference should be same. If you agree i can make both optimized and reference same, and remove optimized ops. So that there will be only reference ops code.

Reply:
Agreed. That is how we usually do.

Reply:
Can we make compute_slice a helper function rather than std::function? I think we generally try to avoid using std::function in our reference ops, as they're shared with TFLite Micro.